tardis/apps/search/utils/documents.py

```diff
@@ -0,0 +1,207 @@
+    DatasetParameter,
+    DatafileParameter,
+)
+
+from tardis.apps.projects.models import (
+    ProjectParameter,
+)
+
+
+def generic_acl_structure():
+    """
+    Return the ES structure of an ACL.
+
+    - pluginId = type of ACL owner: user/group/token
+    - entityId = ID of the owner
+    """
+    return fields.NestedField(
+        properties={
+            "pluginId": fields.KeywordField(),
+            "entityId": fields.KeywordField(),
+            "canDownload": fields.BooleanField(),
+            "canSensitive": fields.BooleanField(),
+        }
+    )
+
+
+def generic_parameter_structure():
+    """
+    Return the ES structure of object parameters and schema.
+    The parameter structure splits out string/numerical/datetime
+    parameters so that ES can specifically handle each of their
+    datatypes.
+
+    - Schemas:
+      - schema_id: Id of the object schemas
+    - string/numerical/datetime:
+      - pn_id: Id of parameter name
+      - pn_name: Name of parameter name
+      - value: value of parameter
+      - sensitive: whether parameter name is sensitive
+    """
+    return fields.NestedField(
+        properties={
+            "string": fields.NestedField(
+                properties={
+                    "pn_id": fields.KeywordField(),
+                    "pn_name": fields.KeywordField(),
+                    "value": fields.TextField(),
+                    "sensitive": fields.BooleanField(),
+                }
+            ),
+            "numerical": fields.NestedField(
+                properties={
+                    "pn_id": fields.KeywordField(),
+                    "pn_name": fields.KeywordField(),
+                    "value": fields.FloatField(),
+                    "sensitive": fields.BooleanField(),
+                }
+            ),
+            "datetime": fields.NestedField(
+                properties={
+                    "pn_id": fields.KeywordField(),
+                    "pn_name": fields.KeywordField(),
+                    "value": fields.DateField(),
+                    "sensitive": fields.BooleanField(),
+                }
+            ),
+            "schemas": fields.NestedField(
+                properties={"schema_id": fields.KeywordField()}
+            ),
+        },
+    )
+
+
+def prepare_generic_acls_build(INSTANCE_ACL_SET, return_list):
+    """Returns the ACLs associated with this
+    object, formatted for elasticsearch.
+    """
+    for acl in INSTANCE_ACL_SET:
+        acl_dict = {}
+        if acl["user__id"] is not None:
+            acl_dict["pluginId"] = "django_user"
+            acl_dict["entityId"] = acl["user__id"]
+        if acl["group__id"] is not None:
+            acl_dict["pluginId"] = "django_group"
+            acl_dict["entityId"] = acl["group__id"]
+        if acl["token__id"] is not None:
+            # token access shouldn't be added to search
+            # unless search is given a way of checking token expiry
+            continue
```

Comment:
Should this return None or something empty at this point? If I follow the code correctly this will return an acl dictionary without a pluginId or entityId, but will pick up the canDownload/canSensitive.

Reply:
Nvm - just realised my noob error 😁